### PR TITLE
[Topi, ARM] Disbale Winograd for quantized tensors.

### PR DESCRIPTION
**Pull request by anijain2305**
Winograd Conv2D on integer tensors lead to bad accuracy. I think this is due to underlying maths that needs floating point multiplication or high precision computation. Enabling it only for FP32 tensors.

@masahi @FrozenGene @zhen-jia @u99127 

Without winograd, the TOP 5 labels for few images with class 0 label is 
~~~
1 [  0 389 391 758 394]
2 [390 394  48 103   0]
3 [  0 389 241 240  28]
4 [133 124 132 131 554]
5 [  0 389 390 939 391]
6 [  0 389  37 997 241]
7 [  0 758 389 391 696]
8 [  0 389 838 930 393]
9 [  0 389   1 395 758]
10 [  0 391 758 389 921]
11 [  0 389   1 391 395]
12 [  0 241 955 944 240]
13 [  0 389   5 393  57]
14 [758 391 346 268 395]
15 [  0 389   1 838 393]
16 [  0 389 241 838  28]
17 [  0 389   1 395  37]
~~~

With winograd, the TOP 5 labels for few images with class 0 label is 
~~~
1 [904 818 828 556 648]
2 [904 632 482 270 556]
3 [904 556 270 632 650]
4 [904 828 650 556 383]
5 [904 632 270 556 482]
6 [904 650 828 556 270]
7 [904 818 650 270 249]
8 [904 632 818  60 482]
9 [904 632 828 556 270]
10 [904 818 556 650 270]
11 [904 632 482 556 270]
12 [904 828 632 482 270]
13 [904 632 818 270 556]
14 [904 828 650 818 556]
15 [904 632 482  60 818]
~~~
Essentially, accuracy goes down to 0%

### CODE REVIEW

**Comment by masahi** [comment]
Do you have concrete data? Say accuracy difference on imagenet with and without winograd.

**Comment by anijain2305** [comment]
Output is all zero. It is not doing the right compute anymore, so it 0% accuracy.

**Comment by anijain2305** [comment]
It might be because of this 
https://github.com/apache/incubator-tvm/blob/c936a81dab2b4b4b595d02153a6654b9d4e09cd5/topi/python/topi/nn/winograd_util.py#L83-L131

The numbers here need to FP32

**Comment by masahi** [comment]
I will take a look at how MKLDNN does winograd on int8.

I also wonder if weight quantization is working correctly. Since transformed weight shape is not in  (OC, OI, KH, KW) format, maybe per channel quantize is broken if winograd is not taken into acount.

**Comment by masahi** [comment]
Seems they cast inputs to fp32 and does multiply in fp32 https://github.com/oneapi-src/oneDNN/blob/2994eae18d84d4c7eab2f1402444b91bffa140a9/src/cpu/jit_avx512_core_u8s8s32x_wino_convolution.cpp#L149-L153

Transform coefficients they use https://github.com/oneapi-src/oneDNN/blob/2994eae18d84d4c7eab2f1402444b91bffa140a9/src/cpu/jit_avx512_core_u8s8s32x_wino_convolution.cpp#L43-L44

**Comment by anijain2305** [comment]
Thanks for sharing the pointers @masahi 
That makes sense. It also explains why accuracy is bad in master. We do not cast back to FP32.

Maybe we can handle that in a separate PR.

**Comment by masahi** [comment]
If the only missing bits is the cast to fp32, I hope our perf result is still valid...

File: python/tvm/relay/op/strategy/arm_cpu.py [hunk 1]
```diff
@@ -59,16 +59,21 @@ def conv2d_strategy_arm_cpu(attrs, inputs, out_type, target):
...
                     wrap_compute_conv2d(topi.x86.conv2d_nchw),
                     wrap_topi_schedule(topi.x86.schedule_conv2d_nchw),
                     name="conv2d_nchw.x86")
+
                 # check if winograd algorithm is applicable
                 _, _, kh, kw = get_const_tuple(kernel.shape)
                 pt, pl, pb, pr = topi.nn.get_pad_tuple(padding, (kh, kw))
-                if kh == 3 and kw == 3 and stride_h == 1 and stride_w == 1 and \
-                    dilation_h == 1 and dilation_w == 1:
+                is_dtype_fp32 = data.dtype == "float32" and kernel.dtype == "float32"
```
Comment (cbalint13) [inline python/tvm/relay/op/strategy/arm_cpu.py:72]: Float16 Too ? (perhaps double aka float64 too)
I get good results with it (e.g yolov3-tiny on mali).
Wouldn't be better to deny int8/int16/int32 ?

**Comment by cbalint13** [comment]
> It might be because of this
> https://github.com/apache/incubator-tvm/blob/c936a81dab2b4b4b595d02153a6654b9d4e09cd5/topi/python/topi/nn/winograd_util.py#L83-L131
> 
> The numbers here need to FP32

```degree``` is an integer (fits in INT8) that requests filter size (usually 3x3), it will also reflect the degree of polynomial.

**Comment by cbalint13** [comment]
> Thanks for sharing the pointers @masahi
> That makes sense. It also explains why accuracy is bad in master. We do not cast back to FP32.
> 
> Maybe we can handle that in a separate PR.
Casting into FP is just a workaround.
Winograd on INT is doable but is tricky.
One wise approach is: https://arxiv.org/abs/2003.08646

**Comment by cbalint13** [comment]
> Output is all zero. It is not doing the right compute anymore, so it 0% accuracy.

I think operations overflows badly (untested, its just supposition) leading to lots of 0 MUL ops.

**Comment by masahi** [comment]
A related discussion at https://github.com/andravin/wincnn/issues/16

It might be better to do weight transform before weight quantization. But given the order of Relay passes, I don't see how that can be achieved.

**Comment by tqchen** [comment]
We might want to do winograd transformation explicitly in relay(as opposed to implicitly in topi), that would enable quantization after the transformation.

But for this particular PR, perhaps we could go ahead and disable i8 winograd for now

**Comment by anijain2305** [comment]
@cbalint13 @masahi PTAL again. I relaxed to have any type of floats. And also added the Top-5 labels for few images in the desciption.

**Comment by masahi** [comment]
Thanks @anijain2305 @cbalint13 @tqchen

**Comment by chrschinab** [comment]
I currently face the issue that Winograd implementation for quantized int8 operators destroys accuracy (see my [topic](https://discuss.tvm.ai/t/tvm-auto-tuning-winograd-conv2d-implementation-for-int8-gives-wrong-classification-results/7182) on discuss.tvm.ai) as described above.

Is the functionality to disable Winograd for quantized operators already implemented? How to do that?

**Comment by BUG1989** [comment]
@masahi  In our project. The int8 winograd F(2,3) has the same accuracy as original int8 conv3x3s1, please refer it.
https://github.com/andravin/wincnn/issues/16#issuecomment-467704201
[ncnn int8 pr](https://github.com/Tencent/ncnn/pull/749)
[naive c code](https://github.com/BUG1989/ncnn/blob/ncnn-pr/src/layer/x86/convolution_3x3_int8.h)

**Comment by cbalint13** [comment]
@BUG1989 ,

Thanks for the share and links ! I was aware of that work, will try to add a schedule to TVM for int8 cases.

**Pull request merged**